### PR TITLE
[OPIK-2858] [BE/FE] Fix playground models not receiving context for base64 images

### PR DESCRIPTION
**Pull request by YarivHashaiComet**
## Details
Root cause: Vision-capable models weren't receiving properly structured content when images were sent as base64 or via datasets.

### Backend Changes (MessageContentNormalizer)
- Parse `<<<image>>>...<<</image>>>` tags and expand to structured content for vision models
- Added expandImagePlaceholders() and expandUserMessage() methods
- Unescape HTML entities for backward compatibility with Mustache templates

### Frontend Changes (usePromptDatasetItemCombination)
- Added wrapImageUrlsWithTags() to detect and wrap raw base64/HTTP image URLs
- Wraps detected images with `<<<image>>>...<<</image>>>` tags for backend processing
- **Refactored to apply DRY principle**: Created reusable helper function to eliminate code duplication
- **Improved maintainability**: Now uses `IMAGE_TAG_START` and `IMAGE_TAG_END` constants from `@/lib/llm` instead of hardcoded strings
- **Eliminated regex duplication**: Exported `DATA_IMAGE_REGEX` and `IMAGE_URL_REGEX` from `images.ts` and reused them instead of defining duplicate patterns
- **Better coverage**: Shared regex patterns support 30+ image formats (vs original 11), including apng, avif, heic, heif, psd, raw, and more
- **Fixed logic bug**: Corrected duplicate URL wrapping by checking context at specific match position instead of searching entire string
- **Performance improvement**: Changed from O(n*m) to O(1) complexity per match by using offset-based context checking

## Change checklist
- [x] User facing
- [ ] Documentation update

## Issues
- Resolves
- 🔗  **[OPIK-2858](https://comet-ml.atlassian.net/browse/OPIK-2858)**

## Testing
- Added 8 backend tests covering single/multiple images, base64, HTML unescaping, edge cases
- All backend tests passing
- Manual testing: GPT-5 successfully analyzed 10 base64 images from dataset
- Frontend: All images.ts unit tests passing (19 tests)
- TypeScript compilation successful

## Documentation
N/A - Internal bug fix, no new configuration or user-facing changes

[OPIK-2858]: https://comet-ml.atlassian.net/browse/OPIK-2858?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by CometActions** [comment]
✅ **Test environment is now available!**

### Access Information
- **URL:** https://pr-3839.dev.comet.com
- **Cluster:** comet-ml-development
- **Namespace:** pr-3839
- **Version:** 1.8.91.3839-merge-259
- **Application logs:** [View in Grafana](https://grafana.dev.comet.com/explore?schemaVersion=1&panes=%7B%22ai4%22%3A%7B%22datasource%22%3A%22P8E80F9AEF21F6940%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22expr%22%3A%22%7Bnamespace%3D%5C%22pr-3839%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22P8E80F9AEF21F6940%22%7D%2C%22editorMode%22%3A%22builder%22%2C%22direction%22%3A%22forward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1)

The deployment has completed successfully and the version has been verified.

File: apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputActions/usePromptDatasetItemCombination.ts [hunk 1]
```diff
@@ -50,20 +50,66 @@ const transformMessageIntoProviderMessage = (
...
+ * - data:image/...;base64,... (base64 encoded images)
+ * - http(s)://... image URLs
+ * - [image_N] placeholders (from processInputData)
+ */
+const wrapImageUrlsWithTags = (content: string): string => {
+  let processedContent = content;
+
+  // Pattern 1: Match data:image base64 strings
+  // Captures the full data URL including the base64 data
+  const DATA_IMAGE_REGEX = /data:image\/[^;]+;base64,[A-Za-z0-9+/]+=*/g;
```
Comment (JetoPistola) [inline apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputActions/usePromptDatasetItemCombination.ts]: In images.ts, there are regexes to use, to follow DRY principle

File: apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputActions/usePromptDatasetItemCombination.ts [hunk 2]
```diff
@@ -50,20 +50,66 @@ const transformMessageIntoProviderMessage = (
     throw new Error(`${notDefinedVariables.join(", ")} not defined`);
   }
 
+  let renderedContent = mustache.render(
```
Comment (JetoPistola) [inline apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputActions/usePromptDatasetItemCombination.ts]: NIT: const, rather than let

File: apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputActions/usePromptDatasetItemCombination.ts [hunk 3]
```diff
@@ -50,20 +50,66 @@ const transformMessageIntoProviderMessage = (
...
+
+  // Pattern 2: Match http(s) image URLs
+  // Only match URLs that end with common image extensions or contain image in path
+  const HTTP_IMAGE_REGEX =
+    /https?:\/\/[^\s<>"{}|\\^`\]]+\.(?:jpg|jpeg|png|gif|webp|bmp|svg|ico|tiff|tif|heic|heif)(?:\?[^\s<>"{}|\\^`\]]*)?(?:#[^\s<>"{}|\\^`\]]*)?/gi;
+
+  // First, wrap data URLs
+  processedContent = processedContent.replace(DATA_IMAGE_REGEX, (match) => {
+    // Check if already wrapped
+    if (processedContent.includes(`<<<image>>>${match}<<</image>>>`)) {
```
Comment (JetoPistola) [inline apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputActions/usePromptDatasetItemCombination.ts]: Use:

```
export const IMAGE_TAG_START = "<<<image>>>";
export const IMAGE_TAG_END = "<<</image>>>";
```

File: apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputActions/usePromptDatasetItemCombination.ts [hunk 4]
```diff
@@ -50,20 +50,66 @@ const transformMessageIntoProviderMessage = (
...
+      return match;
+    }
+    return `<<<image>>>${match}<<</image>>>`;
```
Comment (JetoPistola) [inline apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputActions/usePromptDatasetItemCombination.ts]: Use:

```
export const IMAGE_TAG_START = "<<<image>>>";
export const IMAGE_TAG_END = "<<</image>>>";
```

And calculate the special string, once, to make it 2 rows, like:
const a = b;
return c.includes(a) ? match : a

File: apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputActions/usePromptDatasetItemCombination.ts [hunk 5]
```diff
@@ -50,20 +50,66 @@ const transformMessageIntoProviderMessage = (
...
+  });
+
+  // Then, wrap HTTP(S) image URLs
+  processedContent = processedContent.replace(HTTP_IMAGE_REGEX, (match) => {
```
Comment (JetoPistola) [inline apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputActions/usePromptDatasetItemCombination.ts]: Feels like duplication, same logic like above

**Comment by CometActions** [comment]
✅ **Test environment is now available!**

### Access Information
- **URL:** https://pr-3839.dev.comet.com
- **Cluster:** comet-ml-development
- **Namespace:** pr-3839
- **Version:** 1.8.93.3839-merge-264
- **Application logs:** [View in Grafana](https://grafana.dev.comet.com/explore?schemaVersion=1&panes=%7B%22ai4%22%3A%7B%22datasource%22%3A%22P8E80F9AEF21F6940%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22expr%22%3A%22%7Bnamespace%3D%5C%22pr-3839%5C%22%7D%22%2C%22queryType%22%3A%22range%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22P8E80F9AEF21F6940%22%7D%2C%22editorMode%22%3A%22builder%22%2C%22direction%22%3A%22forward%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%7D%7D&orgId=1)

The deployment has completed successfully and the version has been verified.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/llm/MessageContentNormalizer.java [hunk 6]
```diff
@@ -2,39 +2,47 @@
...
 import java.util.List;
+import java.util.regex.Pattern;
 
 @Slf4j
 @UtilityClass
 public class MessageContentNormalizer {
 
     public static final String IMAGE_PLACEHOLDER_START = "<<<image>>>";
     public static final String IMAGE_PLACEHOLDER_END = "<<</image>>>";
+    private static final Pattern IMAGE_PLACEHOLDER_PATTERN = Pattern.compile(
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/llm/MessageContentNormalizer.java:26]: This constant already exists. See `IMAGE_PLACEHOLDER_PATTERN` in `OnlineScoringEngine`. Please leave a single public reference and re-use it everywhere.

Not a blocker, but desirable to fix in a follow-up change.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/llm/MessageContentNormalizer.java [hunk 7]
```diff
@@ -57,6 +65,103 @@ private ChatCompletionRequest normalizeRequest(ChatCompletionRequest request, bo
...
+        }
+
+        var expandedMessages = new ArrayList<Message>(request.messages().size());
+        for (var message : request.messages()) {
+            if (message instanceof UserMessage userMessage && userMessage.content() instanceof String content) {
+                expandedMessages.add(expandUserMessage(userMessage, content));
+            } else {
+                expandedMessages.add(message);
+            }
+        }
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/llm/MessageContentNormalizer.java:89]: Minor: I'd do this in the stream above, instead of iterating again here.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/llm/MessageContentNormalizer.java [hunk 8]
```diff
@@ -57,6 +65,103 @@ private ChatCompletionRequest normalizeRequest(ChatCompletionRequest request, bo
...
+     * Expands a UserMessage with string content containing image placeholders to structured content.
+     * Parses image placeholders and creates a list of Content objects (text and image_url).
+     */
+    private UserMessage expandUserMessage(UserMessage userMessage, String content) {
+        var matcher = IMAGE_PLACEHOLDER_PATTERN.matcher(content);
+
+        if (!matcher.find()) {
+            // No image placeholders found, return as-is
+            return userMessage;
+        }
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/llm/MessageContentNormalizer.java:106]: Minor: This overlaps a bit with the `needsExpansion` logic above. I'd centralise in one place.

**Review by andrescrz — APPROVED** [review]
Just a few recommendations, but LGTM.

**Review by aadereiko — APPROVED** [review]
I haven't tested it, but the code itself looks ok
